src/bun.js/api/glob.zig

```diff
@@ -0,0 +1,381 @@
+                    globalThis.throwOutOfMemory();
+                    return null;
+                };
+            };
+
+            if (cwd_str.len > bun.MAX_PATH_BYTES) {
+                globalThis.throw("{s}: invalid `cwd`, longer than {d} bytes", .{ fnName, bun.MAX_PATH_BYTES });
+                return null;
+            }
+
+            out.cwd = BunString.fromBytes(cwd_str);
+        }
+
+        if (optsObj.get(globalThis, "dot")) |dot| {
+            out.dot = if (dot.isBoolean()) dot.asBoolean() else false;
+        }
+
+        return out;
+    }
+};
+
+pub const WalkTask = struct {
+    walker: *GlobWalker,
+    alloc: Allocator,
+    err: ?Err = null,
+    global: *JSC.JSGlobalObject,
+    pub const Err = union(enum) {
+        syscall: Syscall.Error,
+        unknown: anyerror,
+
+        pub fn toJSC(this: Err, globalThis: *JSGlobalObject) JSValue {
+            return switch (this) {
+                .syscall => |err| err.toJSC(globalThis),
+                .unknown => |err| ZigString.fromBytes(@errorName(err)).toValueGC(globalThis),
+            };
+        }
+    };
+
+    pub const AsyncGlobWalkTask = JSC.ConcurrentPromiseTask(WalkTask);
+
+    pub fn create(globalThis: *JSC.JSGlobalObject, alloc: Allocator, globWalker: *GlobWalker) !*AsyncGlobWalkTask {
+        var walkTask = try alloc.create(WalkTask);
+        walkTask.* = .{
+            .walker = globWalker,
+            .global = globalThis,
+            .alloc = alloc,
+        };
+        return try AsyncGlobWalkTask.createOnJSThread(alloc, globalThis, walkTask);
+    }
+
+    pub fn run(this: *WalkTask) void {
+        const result = this.walker.walk() catch |err| {
+            this.err = .{ .unknown = err };
+            return;
+        };
+        switch (result) {
+            .err => |err| {
+                this.err = .{ .syscall = err };
+            },
+            .result => {},
+        }
+    }
+
+    pub fn then(this: *WalkTask, promise: *JSC.JSPromise) void {
+        defer this.deinit();
+
+        if (this.err) |err| {
+            const errJs = err.toJSC(this.global);
+            promise.reject(this.global, errJs);
+            return;
+        }
+
+        const jsStrings = globWalkResultToJS(this.walker, this.global);
+        promise.resolve(this.global, jsStrings);
+    }
+
+    fn deinit(this: *WalkTask) void {
+        this.walker.deinit();
+        this.alloc.destroy(this);
+    }
+};
+
+fn globWalkResultToJS(globWalk: *GlobWalker, globalThis: *JSGlobalObject) JSValue {
+    // if (globWalk.matchedPaths.items.len >= 0) {
+    if (globWalk.matchedPaths.items.len == 0) {
+        return JSC.JSArray.from(globalThis, &[_]JSC.JSValue{});
+    }
+
+    return BunString.toJSArray(globalThis, globWalk.matchedPaths.items[0..]);
+}
+
+fn makeGlobWalker(
+    this: *Glob,
+    globalThis: *JSGlobalObject,
+    arguments: *ArgumentsSlice,
+    comptime fnName: []const u8,
+    alloc: Allocator,
+    arena_: Arena,
```

Comment:
The pointer to the arena will no longer be valid outside this scope. This is a very confusing thing about the arena allocator in Zig’s standard library.

Reply:
What is the best way to handle this, because the GlobWalker wants to own an `Arena`. Should I pass an `*Arena` to this function and then dereference it to give it to the `GlobWalker`?

Or should `GlobWalker` take an `*Arena`, and only when it is done allocating it can dereference it and store it?